File: Dynamic_Programming/2320.Count_Number_of_Ways_to_Place_Houses/C++/2320.Count_Number_of_Ways_to_Place_Houses.cpp
class Solution {
public:
    int countHousePlacements(int n) 
    {
        //===============================//
        // Input type:                   //
        //   - int (n-th amount)         //
        // Return type:                  //
        //   - int (result combinations) //
        //===============================//

        /*Initialize*/
        ///// Record dynamic-programming (i.e. DP) vector /////
        vector<int> record_dp_vect((++n), 0);

        record_dp_vect[0] = 1;
        record_dp_vect[1] = 2;

        ///// Record module /////
        int record_module = (pow(10, 9) + 7);


        /*Dynamic-programming (i.e. DP) based loop traversal*/
        for (int n_idx = 2; (n_idx < n); (n_idx++)) //Whole
        {
            record_dp_vect[n_idx] += (int)(((long long)(record_dp_vect[(n_idx - 1)]) + (long long)(record_dp_vect[(n_idx - 2)])) % record_module); //Keep updating, overwriting

        } //Whole

        return (int)(((long long)(record_dp_vect[(--n)]) * (long long)(record_dp_vect[n])) % record_module);
    }
};

File: Dynamic_Programming/2320.Count_Number_of_Ways_to_Place_Houses/Python/2320.Count_Number_of_Ways_to_Place_Houses.py
class Solution(object):
    def countHousePlacements(self, n):
        """
        :type n: int
        :rtype: int
        """
        #============================================#
        # Dynamic-programming based traversal method #
        #============================================#

        ############
        #Initialize
        ##### N-th amount #####
        n += 1

        ##### Record dynamic-programming (i.e. DP) array #####
        record_dp_arry = [0 for _ in range(n)]

        record_dp_arry[0] = 1
        record_dp_arry[1] = 2

        ##### Record module ######
        record_module = ((10 ** 9) + 7)


        ##########################################
        #Dynamic-programming based loop traversal
        for n_idx in range(2, n):
            record_dp_arry[n_idx] += ((record_dp_arry[(n_idx - 1)] + record_dp_arry[(n_idx - 2)]) % record_module) #Keep updating/overwriting

        return ((record_dp_arry[(n - 1)] ** 2) % record_module)
